http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/request/CrtRequestAdapter.java

```diff
@@ -131,11 +134,22 @@ private static List<HttpHeader> createHttpHeaderList(URI uri, HttpExecuteRequest
             crtHeaderList.add(new HttpHeader(Header.CONNECTION, Header.KEEP_ALIVE_VALUE));
         }
 
-        // We assume content length was set by the caller if a stream was present, so don't set it here.
+        if (!sdkRequest.firstMatchingHeader(Header.CONTENT_LENGTH).isPresent()
+            && sdkExecuteRequest.contentStreamProvider().isPresent()) {
+            String contentLength = String.valueOf(calculateContentLength(sdkExecuteRequest.contentStreamProvider().get()));
+            crtHeaderList.add(new HttpHeader(Header.CONTENT_LENGTH, contentLength));
+        }
 
-        // Add the rest of the Headers
         sdkRequest.forEachHeader((key, value) -> value.stream().map(val -> new HttpHeader(key, val)).forEach(crtHeaderList::add));
 
         return crtHeaderList;
     }
+
+    public static long calculateContentLength(ContentStreamProvider contentStreamProvider) {
+        try {
+            return  IoUtils.toByteArray(contentStreamProvider.newStream()).length ;
```

Comment:
Wont that buffer the whole body content in memory? Do we really want to do that?

Reply:
Need to rework on this , due to internal comment from Matthew miller to handle this in Marshallers.  Will update the PR with that approach.